itemadapter/_json_schema.py

```diff
@@ -0,0 +1,579 @@
+            prop.setdefault("type", "object")
+            args = get_args(prop_type)
+            if args:
+                assert len(args) == 2
+                value_type = args[1]
+                props = prop.setdefault("additionalProperties", {})
+                update_prop_from_type(props, value_type, state)
+            return
+    if origin in UNION_TYPES:
+        update_prop_from_union(prop, prop_type, state)
+
+
+def update_prop_from_type(prop: dict[str, Any], prop_type: Any, state: _JsonSchemaState) -> None:
+    if (origin := get_origin(prop_type)) is not None:
+        update_prop_from_origin(prop, origin, prop_type, state)
+        return
+    if isinstance(prop_type, type):
+        if state.adapter.is_item_class(prop_type):
+            if prop_type in state.containers:
+                prop.setdefault("type", "object")
+                return
+            state.containers.add(prop_type)
+            subschema = state.adapter.get_json_schema(
+                prop_type,
+                _state=state,
+            )
+            state.containers.remove(prop_type)
+            for k, v in subschema.items():
+                prop.setdefault(k, v)
+            return
+        if issubclass(prop_type, Enum):
+            values = [item.value for item in prop_type]
+            prop.setdefault("enum", values)
+            value_types = tuple({type(v) for v in values})
+            prop_type = value_types[0] if len(value_types) == 1 else Union[value_types]
+            update_prop_from_type(prop, prop_type, state)
+            return
+        if not issubclass(prop_type, str):
+            if isinstance(prop_type, ObjectProtocol):
+                prop.setdefault("type", "object")
+                return
+            if isinstance(prop_type, ArrayProtocol):
+                prop.setdefault("type", "array")
+                if issubclass(prop_type, AbstractSet):
+                    prop.setdefault("uniqueItems", True)
+                return
+    json_schema_type = SIMPLE_TYPES.get(prop_type)
+    if json_schema_type is not None:
+        prop.setdefault("type", json_schema_type)
+
+
+def _setdefault_attribute_types_on_json_schema(
+    schema: dict[str, Any], item_class: type, state: _JsonSchemaState
+) -> None:
+    """Inspect the type hints of the class attributes of the item class and,
+    for any matching JSON Schema property that has no type set, set the type
+    based on the type hint."""
+    props = schema.get("properties", {})
+    attribute_type_hints = get_type_hints(item_class)
+    for prop_name, prop in props.items():
+        if prop_name not in attribute_type_hints:
+            continue
+        prop_type = attribute_type_hints[prop_name]
+        update_prop_from_type(prop, prop_type, state)
+
+
+def _setdefault_attribute_docstrings_on_json_schema(
+    schema: dict[str, Any], item_class: type
+) -> None:
+    """Inspect the docstrings after each class attribute of the item class and,
```

Comment:
Would it make sense to move the "inspect the docstring after each class attribute" part to its own function?